design/census-import/generate-census-records.js

```diff
@@ -26,53 +26,76 @@ function writeCsv (data) {
+  FROM ${sqlService.adminSchema}.school`
+  const result = await sqlService.query(sql, [])
+  return R.prop('cnt', R.head(result))
+}
+
+function generateUpn (dfeNumber) {
+  const leaCode = dfeNumber.slice(0, 3)
+  const estabCode = dfeNumber.slice(3, dfeNumber.length)
+  const minYearValue = academicYear - 11
+  const maxYearValue = academicYear - 7
+  const year = Math.floor(Math.random() * (maxYearValue - minYearValue + 1) + minYearValue).toString().substr(-2)
+  let baseUpn = leaCode + estabCode + year
+  pupilCount += 1
+  if (pupilCount > 999) {
+    pupilCount = 1
+    baseUpn = (parseInt(baseUpn) + 1000).toString()
+  }
+  const serial = pupilCount.toString().padStart(3, '0')
+  return upnService.calculateCheckLetter(baseUpn + serial) + baseUpn + serial
 }
 
-async function main (options) {
-  winston.info('Generating census records...')
-  const { recordsLength } = options
+async function main () {
+  await sqlService.initPool()
+  console.log('Generating census records...')
   const csvData = []
-  let baseUpn = '702500001'
-  let schoolsLength
+  let dfeNumbers
+  let dfeNumberCount
   try {
-    schoolsLength = await getSchoolsLength()
-    schoolsLength = schoolsLength - 1
+    dfeNumbers = await getSchoolDfeNumbers()
+    dfeNumberCount = await getSchoolDfeNumberCount()
+    dfeNumberCount = dfeNumberCount - 1
   } catch (error) {
-    winston.info(error)
+    console.log(error)
   }
-  for (let i = 0; i <= recordsLength; i++) {
-    let pupilIdx = i + 1
-    if (pupilIdx > 999) {
-      baseUpn = (parseInt(baseUpn) + 1000).toString()
-      pupilIdx = 1
+  for (let i = 0; i <= dfeNumberCount; i++) {
+    for (let j = 0; j <= 31; j++) {
+      const dfeNumber = dfeNumbers[i].toString()
+      const upn = generateUpn(dfeNumber, pupilCount)
+      const record = [
+        dfeNumber.slice(0, 3), // LEA
+        dfeNumber.slice(3, dfeNumber.length), // Estab
+        upn,
+        faker.name.lastName(), // Surname
+        faker.name.firstName(), // Forename
+        i % 100 === 0 ? faker.name.firstName() : '', // Middlenames
+        i % 10 === 0 ? 'M' : 'F', // Gender
+        momentRandom(
+          moment.utc(`${academicYear - 7}-09-01`),
+          moment.utc(`${academicYear - 11}-09-02`)
+        ).format('DD/MM/YYYY') // DOB
+      ]
+      csvData.push(record)
     }
-    const serial = pupilIdx.toString().padStart(3, '0')
-    const upn = upnService.calculateCheckLetter(baseUpn + serial) + baseUpn + serial
-    const randomEstabCode = (Math.floor(Math.random() * schoolsLength) + 1001).toString()
-    const record = [
-      '999', // LEA
-      randomEstabCode, // Estab
-      upn,
-      faker.name.lastName(), // Surname
-      faker.name.firstName(), // Forename
-      i % 100 === 0 ? faker.name.firstName() : '', // Middlenames
-      i % 10 === 0 ? 'M' : 'F', // Gender
-      moment(faker.date.between('2005', '2006')).format('MM/DD/YY') // DOB
-    ]
-    csvData.push(record)
   }
   writeCsv(csvData)
```

Comment:
Is `writeCSV()` being called only once? i.e. the whole lot is stored in memory.  If so, what is the memory required for this process?

Reply:
Okay, could be worse 😀 

That's fine with me for a one-off dev tool